### PR TITLE
fix: prevent extracting archived files outside of target path

### PR DESCRIPTION
**Pull request by aviadatsnyk**
## Why this PR?
This PR is meant to fix an arbitrary file write vulnerability, that can be achieved using a specially crafted zip archive, that holds path traversal filenames. When the filename gets concatenated to the target extraction directory, the final path ends up outside of the target folder. 

A sample malicious zip file named zip-slip.zip. (see [this gist](https://gist.github.com/grnd/2992cc4d968386ae277123bad9e7f00a)) was used, and when running the code below, resulted in creation of `evil.txt` file in /tmp folder.

```js
fs.createReadStream('zip-slip.zip').pipe(unzipper.Extract({ path: '/tmp/safe' }));
```

There are various possible ways to avoid this issue, some include checking for `..` (dot dot) characters in the filename, but the best solution in our opinion is to check if the final target filename, starts with the target folder (after both are resolved to their absolute path).

Stay secure,
Snyk Team

### CODE REVIEW

File: lib/extract.js [hunk 1]
```diff
@@ -17,8 +17,17 @@ function Extract (opts) {
 
   self.on('entry', function(entry) {
     if (entry.type == 'Directory') return;
+
+    // to avoid zip slip (writing outside of the destination), we resolve
+    // the target path, and make sure it's nested in the intended
+    // destination, or not extract it otherwise.
+    var extractPath = path.join(opts.path, entry.path);
+    if (extractPath.indexOf(opts.path) != 0) {
+      return;
```
Comment (ZJONSSON) [inline lib/extract.js]: This should either throw an error or at the very least show a warning.
Also the `entry` is left hanging here as nothing drains it  (possibly use .autodrain)

**Comment by ZJONSSON** [comment]
published as `unzipper@0.8.13`.  Awesome catch, thank you!!

**Comment by glen-nicol** [comment]
Hi there, I ended up using the code snippet above to add zip-slip protection to some of my manual unzipping methods that use this library. And I noticed that on my linux docker image it failed to detect the [windows zip-slip archive](https://github.com/snyk/zip-slip-vulnerability/blob/master/archives/zip-slip-win.zip).

I am using the following and it catches both of the examples from synk on windows and linux docker images (did not try mac)

```javascript
const DUMMY_PATH = path.join('/sandbox');
const extractPath = path.join(DUMMY_PATH, zipEntry.path.replace(/\\/g, '/'));
if (extractPath.indexOf(DUMMY_PATH) !== 0) {
  return true;
}
```

**Comment by ZJONSSON** [comment]
Good catch, can you do a PR with a test that fails without the fix above?

**Comment by glen-nicol** [comment]
Sure, I'll try to do that in the next couple days.

**Comment by ZJONSSON** [comment]
Awesome, thank you @glen-nicol

**Comment by glen-nicol** [comment]
I was looking at it some more, I don't think UNIX machines are actually vulnerable to the windows zip. But I think it will end up creating a rather strange filename "................\tempevil.txt". Should we still go ahead and make the change to ignore something that is obviously attempting to do something bad?

**Comment by ZJONSSON** [comment]
I ran tests against the  [windows zip-slip archive](https://github.com/snyk/zip-slip-vulnerability/blob/master/archives/zip-slip-win.zip) on a windows machine and the file was safely ignored there.    Ideally we should want same behavior independent of platform so the preference here would be to ignore.      But I'm happy that this is not actual vulnerability

**Pull request merged**